### PR TITLE
Java Backend Logging: First batch of logging features

### PR DESCRIPTION
**Pull request by denis-bogdanas**
## Main additions
- option --log to print a subset of configuration specified in `--log-cells` at each step.
- option --log-basic - same as --log but only at certain important steps such as branchings, application of spec rule.
  Plus enables multiple other minor logs.
- option --log-cells. Generalized logging. Added ability to specify the list of cells to log, and which of these cells to be pretty-printed.
- option --branching-allowed to limit branchings. If allowed number of branchings is exceeded, execution ends.
- option --log-rule, prints rule source for all invoked rules of in all contexts. Both kprove and krun.
- final implication debug: Logging the non-matching sub-terms when final implication does not match, if at least --log-basic is provided.

## Minor additions
- added parameter `step` to some methods, for debugging.
- Custom formatting when entire term is printed, without using the slow pretty printing.

### CODE REVIEW

**Comment by denis-bogdanas** [comment]
@ehildenb ready for review.

File: java-backend/src/main/java/org/kframework/backend/java/symbolic/JavaExecutionOptions.java [hunk 1]
```diff
@@ -54,6 +55,37 @@
             description="Clear function cache after initialization phase. Frees some memory. Use IN ADDITION to --cache-func")
     public boolean cacheFunctionsOptimized = false;
 
+    @Parameter(names="--branching-allowed", arity=1, description="Number of branching events allowed before a forcible stop.")
+    public int branchingAllowed = Integer.MAX_VALUE;
+
+    @Parameter(names="--log", description="Log every step. KEVM only.")
```
Comment (ehildenb) [inline java-backend/src/main/java/org/kframework/backend/java/symbolic/JavaExecutionOptions.java]: This isn't accurate anymore, AFAICT.

File: java-backend/src/main/java/org/kframework/backend/java/symbolic/JavaExecutionOptions.java [hunk 2]
```diff
@@ -54,6 +55,37 @@
...
+    public boolean log = false;
+
+    @Parameter(names="--log-basic",
+            description="Log most basic information: summary of initial step, final steps and final implications." +
+                    " All custom logging only works for KEVM-based specs.")
```
Comment (ehildenb) [inline java-backend/src/main/java/org/kframework/backend/java/symbolic/JavaExecutionOptions.java]: Same here.

**Review by ehildenb — CHANGES_REQUESTED** [review]

**Comment by denis-bogdanas** [comment]
@ehildenb done.

**Review by ehildenb — APPROVED** [review]

**Pull request merged**